From: Borislav Petkov <bp@suse.de>
Date: Sun, 17 Dec 2017 16:37:58 +0100
Subject: x86/CPU: Check speculation control CPUID bit
Patch-mainline: submitted on 2018/1/9
References: bsc#1068032

... and enable the corresponding flags.

Carved out from a patch by Tim Chen <tim.c.chen@linux.intel.com> and
improved.

After microcode reload, we need to check CPUID directly as we don't
update the X86_FEATURE flags after a reload.

Signed-off-by: Borislav Petkov <bp@suse.de>
---
 arch/x86/include/asm/spec_ctrl.h     |  1 +
 arch/x86/kernel/cpu/intel.c          |  3 +++
 arch/x86/kernel/cpu/microcode/core.c |  6 +++++-
 arch/x86/kernel/cpu/spec_ctrl.c      | 16 ++++++++++++++++
 4 files changed, 25 insertions(+), 1 deletion(-)

diff --git a/arch/x86/include/asm/spec_ctrl.h b/arch/x86/include/asm/spec_ctrl.h
index 384b3b8b79b3..96f61a82396d 100644
--- a/arch/x86/include/asm/spec_ctrl.h
+++ b/arch/x86/include/asm/spec_ctrl.h
@@ -96,6 +96,7 @@ void x86_disable_ibrs(void);
 void stuff_RSB(void);
 unsigned int x86_ibrs_enabled(void);
 unsigned int x86_ibpb_enabled(void);
+void x86_spec_check(void);
 
 static inline void x86_ibp_barrier(void)
 {
diff --git a/arch/x86/kernel/cpu/intel.c b/arch/x86/kernel/cpu/intel.c
index 176ccd2964ff..f68e0bc85f2f 100644
--- a/arch/x86/kernel/cpu/intel.c
+++ b/arch/x86/kernel/cpu/intel.c
@@ -16,6 +16,7 @@
 #include <asm/intel-family.h>
 #include <asm/hwcap2.h>
 #include <asm/elf.h>
+#include <asm/spec_ctrl.h>
 
 #ifdef CONFIG_X86_64
 #include <linux/topology.h>
@@ -560,6 +561,8 @@ static void init_intel(struct cpuinfo_x86 *c)
 		detect_vmx_virtcap(c);
 
 	probe_xeon_phi_r3mwait(c);
+
+	x86_spec_check();
 }
 
 #ifdef CONFIG_X86_32
diff --git a/arch/x86/kernel/cpu/microcode/core.c b/arch/x86/kernel/cpu/microcode/core.c
index b3e94ef461fd..b7dc3ca4518a 100644
--- a/arch/x86/kernel/cpu/microcode/core.c
+++ b/arch/x86/kernel/cpu/microcode/core.c
@@ -39,6 +39,7 @@
 #include <asm/microcode.h>
 #include <asm/processor.h>
 #include <asm/cmdline.h>
+#include <asm/spec_ctrl.h>
 
 #define MICROCODE_VERSION	"2.01"
 
@@ -417,8 +418,11 @@ static ssize_t reload_store(struct device *dev,
 		if (!ret)
 			ret = tmp_ret;
 	}
-	if (!ret)
+	if (!ret) {
 		perf_check_microcode();
+		x86_spec_check();
+	}
+
 	mutex_unlock(&microcode_mutex);
 	put_online_cpus();
 
diff --git a/arch/x86/kernel/cpu/spec_ctrl.c b/arch/x86/kernel/cpu/spec_ctrl.c
index 672c32da1b02..dbf00ab91589 100644
--- a/arch/x86/kernel/cpu/spec_ctrl.c
+++ b/arch/x86/kernel/cpu/spec_ctrl.c
@@ -49,3 +49,19 @@ void stuff_RSB(void)
 	stuff_rsb();
 }
 EXPORT_SYMBOL_GPL(stuff_RSB);
+
+/*
+ * Called after upgrading microcode, check CPUID directly.
+ */
+void x86_spec_check(void)
+{
+	if (boot_cpu_data.x86_vendor == X86_VENDOR_INTEL) {
+		if (cpuid_edx(7) & BIT(26)) {
+			ibrs_state = 1;
+			ibpb_state = 1;
+
+			setup_force_cpu_cap(X86_FEATURE_SPEC_CTRL);
+		}
+	}
+}
+EXPORT_SYMBOL_GPL(x86_spec_check);

